### DISCUSSION TITLE
"error: 'RSA_SSLV23_PADDING' undeclared" while installing Ruby 2.6.5 on Ubunut 24.04 despite explicitly using OpenSSL 1.1

### DISCUSSION THREAD

**Discussion by BrainStone** [rbenv/ruby-build#2381] [Build failures] [answered]
So I'm getting the error `ossl_pkey_rsa.c:877:58: error: 'RSA_SSLV23_PADDING' undeclared (first use in this function); did you mean 'RSA_NO_PADDING'` while trying to install Ruby 2.6.5 (I'm very aware it's ancient, but that's what my current project uses and while efforts are on the way to upgrade it, that won't happen for quite some time) on my new Ubuntu 24.04 machine.

Having looked around the internet, I came across [this SO thread](https://stackoverflow.com/questions/72250611/rsa-sslv23-padding-undeclared-first-use-in-this-function-did-you-mean-rsa), so I followed the guide (I only changed where the lib is installed) and tried to let `rbenv install` know by doing this:

```zsh
export RUBY_CFLAGS="-Wno-deprecated-declarations"
export CONFIGURE_OPTS="--with-open-ssl-dir=$RBENV_ROOT/deps/lib/openssl"
rbenv install 2.6.5
```

However that did not change anything and I'm still getting the error. I checked with pgrep if the flag was actually passed to `./configure`, which it absolutely is.

Attached is my build log:
[ruby-build.20240503154843.260316.log](https://github.com/rbenv/ruby-build/files/15200957/ruby-build.20240503154843.260316.log)

For reference here is how I built OpenSSL 1.1:

```zsh
cd $RBENV_ROOT/deps/src
wget https://www.openssl.org/source/openssl-1.1.1q.tar.gz
tar xvzf openssl-1.1.1q.tar.gz
cd openssl-1.1.1q
mkdir -p $RBENV_ROOT/deps/lib/openssl
./config --prefix=$RBENV_ROOT/deps/lib/openssl
make -j$(nproc)
make install
cp -r /etc/ssl/certs/* $RBENV_ROOT/deps/lib/openssl/ssl/certs
```

**Comment by mislav** [comment, marked as answer]
Hi, are you using a recent version of ruby-build?

> ```shell
> export CONFIGURE_OPTS="--with-open-ssl-dir=$RBENV_ROOT/deps/lib/openssl"
> ```

I see two potential problems here. First is that `--with-openssl-dir` flag was misspelled (there was an extra dash). Second is that it should be `export RUBY_CONFIGURE_OPTS=...` instead of just CONFIGURE_OPTS to ensure that the flags only apply to Ruby and not to compiling any other software.

If you are using an older version of ruby-build, then explicitly setting this might help:
```sh
export PKG_CONFIG_PATH="$RBENV_ROOT/deps/lib/openssl/lib/pkgconfig:$PKG_CONFIG_PATH"
```
Ref. https://github.com/rbenv/ruby-build/discussions/2125#discussioncomment-7516993

Newer versions of ruby-build automatically set this for you to help with compiling Ruby 2.6.

**Reply by BrainStone** [reply]
Ok. Yes that was it. I don't know how that dash ended up in there.

And I was using both `RUBY_CONFIGURE_OPTS` and `CONFIGURE_OPTS`, but since I misspelled both, nothing changed.

I'm using whatever comes bundled with rbenv on Ubuntu 24.04. The only bit of special sauce was openssl 1.1 being compiled on the machine.
